daml-assistant/daml-helper/src/DA/Daml/Helper/Ledger.hs

```diff
@@ -177,17 +177,28 @@ runLedgerUploadDar flags darPathM = do
       getDarPath
   putStrLn $ "Uploading " <> darPath <> " to " <> showHostAndPort args
   bytes <- BS.readFile darPath
-  uploadDarFile args bytes `catch` \(e :: SomeException) -> do
-    putStrLn $ unlines
-      [ "An exception was thrown when during the upload-dar command"
-      , "- " <> show e
-      , "One reason for this to occur is if the size of DAR file being uploaded exceeds the gRPC maximum message size. The default value for this is 4Mb, but it may be increased when the ledger is (re)started. Please check with your ledger operator."
-      ]
-    exitFailure
-  putStrLn "DAR upload succeeded."
-  where
-    uploadDarFile args bytes =
-      runWithLedgerArgs args $ do L.uploadDarFile bytes >>= either fail return
+  result <- uploadDarFile args bytes
+  case result of
+    Left err -> do
+      putStrLn $
+        unlines
+          [ "An exception was thrown during the upload-dar command"
+          , "- " <> show err
+          , "One reason for this to occur is if the size of DAR file being uploaded exceeds the gRPC maximum message size. The default value for this is 4Mb, but it may be increased when the ledger is (re)started. Please check with your ledger operator."
+          ]
+      exitFailure
+    Right () -> putStrLn "DAR upload succeeded."
+
+uploadDarFile :: LedgerArgs -> BS.ByteString -> IO (Either String ())
+uploadDarFile args bytes =
+  case api args of
+    Grpc -> runWithLedgerArgs args $ do L.uploadDarFile bytes
+    HttpJson -> do
+      (i :: Int) <- httpJsonRequest args "POST" "/v1/packages" $ setRequestBodyLBS (BSL.fromStrict bytes)
+      return $
+        if i == 1
+          then Right ()
+          else Left "An error occured. Please check the returned status."
```

Comment:
What is the "returned status"?

Reply:
The returned status will be reported by the http client itself for example: `daml-helper: user error (Request failed with error code Status {statusCode = 401, statusMessage = "Unauthorized"})`

Reply:
I see. Thanks for the explanation.